tycho-its/src/test/java/org/eclipse/tycho/test/buildextension/PomlessTest.java

```diff
@@ -130,6 +132,21 @@ public void testPomlessModel() throws Exception {
 				"aggregator3-pomless");
 
 		assertThat(projectData, is(aMapWithSize(0))); // Ensure no more projects are found
+
+		// Ensure that only those aggregator-directories that are listed as modules in
+		// the parent (and don't have a pom) are scanned for nested modules.
+		List<String> logLines = verifier.loadFile(verifier.getBasedir(), verifier.getLogFileName(), false);
```

Comment:
I don't think this will work, it is not required that the module is defined in the parent, it could be any pom in any directory (just take https://github.com/eclipse-platform/eclipse.platform.swt/blob/master/.github/aggregator.pom as an example)

Reply:
Actually polyglot-maven only searches in those directories listed as modules so it should work as intended.
But I can add an extra case for such non-nested aggregators.
Parent was referring to the module-tree parent not the pom-model parent. I try to make this more clear.

Reply:
No please try it out, create an empty folder and run maven with `mvn -f myEmptyFolder`, even though no one ever defines any modules, the mappings will be asked to locate a pom.

Reply:
> No please try it out, create an empty folder and run maven with `mvn -f myEmptyFolder`, even though no one ever defines any modules, the mappings will be asked to locate a pom.

OK, but in this case you tell Maven/Polyglot to look for a pom(-model) at that location. Then of course it looks it up there. But subsequently it looks only at the directories that are declared as modules in the just read pom-model. If a pom-type model is read from any of those location it looks at the directories listed as modules in that models as well and so on, until the module-aggreagation-tree is fully traversed.

But what it is NOT happening (at least from my knowledge of the polyglot-code and my observations while debugging for this change) is that starting at the build's root directory the entire file-system is scanned for models. That is what I meant when I said that polyglot doesn't scan directories arbitrarily and only those that are necessary. 

> 
> There is no "module-tree parent", it is just the default behavior that the path to the parent is ".." if you don't specify a parent path and tycho assumes you want to use that as a parent, but that do not mean it must define any modules at all.

Sorry that was unclear said from me. `Aggregator` is the more suitable term here. Although it is not accessible from the model, if you treat the aggregator->module relations as graph, then this is a directed ascylic graph (or tree) where one usually speaks of a parent. But yes in the Maven-world that's the aggregator.

Reply:
> But what it is NOT happening (at least from my knowledge of the polyglot-code and my observations while debugging for this change) is that starting at the build's root directory the entire file-system is scanned for models. That is what I meant when I said that polyglot doesn't scan directories arbitrarily and only those that are necessary.

As said, polyglot has **nothing** to do with this, its is **maven-cli** that triggers this and tycho-pomless that supplies some kind of dynamic modules configuration.

If you run maven, then maven will query the current folder (or the file/folder given with `-f`) for a pom (that's where "locate pom" is called), if that is an existing file, it further reads this model (that's where again polyglot is queried) and if this is a "pom" type and contains modules these are also queried and so on, and that's why I said we can't have a state in this, it must be fully defined by a single file/folder and nothing else, it can be called once or several times and in any order, e.g. m2e will call this too to find out if a pomless project is still a maven project, also note, that one can also have json poms and so on and even combine this in any way

What Tycho now does is that it extends the idea of polyglot by dynamically creating a so called "connector-pom" by again query subfolder of folders that have a special name (e.g. plugins) for a pom, again here, any order, any time) and if these have a pom add them to this folders module list (as an alternative you can create a tycho.pom file and maintain the list manually).

But that also means, if you allow *any* name, that *any* folder would be assumed to be a "connector pom folder" and thus you will add *any* folder as long as it somewhere contains something that could be interpreted as a maven project and that's why your code contains special handling for if it is "the first" I guess.

You can try this out with current Tycho by creating a structure like:

`plugins/plugins/plugin/plugins/<a bundle or traditional pom.xml project>`

then create a pom in the root with just `<module>plugins</plugins>` and it will/should traverse the whole sub-directory structure, creating for each one a connector pom.